cranelift/isle/isle/src/serialize.rs

```diff
@@ -0,0 +1,523 @@
+                let (this, rest) = order.split_at_mut(partition_point);
+                order = rest;
+
+                let check = match best {
+                    MatchKind::Constraint(source) => {
+                        self.use_expr(source);
+                        self.add_bindings();
+                        let mut arms = Vec::new();
+
+                        let get_constraint =
+                            |idx: usize| self.rules.rules[idx].get_constraint(source).unwrap();
+                        this.sort_unstable_by_key(|&idx| get_constraint(idx));
+                        for g in group_by_mut(this, |&a, &b| get_constraint(a) == get_constraint(b))
+                        {
+                            let mut child = self.new_block();
+                            // Applying a constraint moves the discriminant from Emitted to Matched, but
+                            // only within the constraint's match arm; later fallthrough cases may need to
+                            // match this discriminant again.
+                            child.set_ready(source, Readiness::Matched);
+
+                            let constraint = get_constraint(g[0]);
+                            let bindings = Vec::from_iter(
+                                constraint
+                                    .bindings_for(source)
+                                    .into_iter()
+                                    .map(|b| child.rules.find_binding(&b)),
+                            );
+
+                            let mut changed = false;
+                            for &binding in bindings.iter() {
+                                if let Some(binding) = binding {
+                                    // Matching a pattern makes its bindings available, and also emits code to bind them.
+                                    child.set_ready(binding, Readiness::Emitted);
+                                    changed = true;
+                                }
+                            }
+
+                            // As an optimization, only propagate availability if we changed any binding's readiness.
+                            if changed {
+                                child.add_bindings();
+                            }
+
+                            let body = child.sort(g);
+
+                            arms.push(MatchArm {
+                                constraint,
+                                bindings,
+                                body,
+                            });
+                        }
+
+                        Condition::Match { source, arms }
+                    }
+
+                    MatchKind::Equal(a, b) => {
+                        self.use_expr(a);
+                        self.use_expr(b);
+                        self.add_bindings();
+
+                        let mut child = self.new_block();
+                        child.set_ready(a, Readiness::Matched);
```

Comment:
Verify that equality constraints are processed in the right order, such that we always have one non-`Matched` input to trigger the evaluation? Consider e.g. `a == b`, `c == d`, `b == c` in order: after the first two, all four are `Matched` but we haven't yet processed `b == c`...